runtime/compiler/env/VMJ9Server.cpp

```diff
@@ -1179,17 +1179,37 @@ TR_J9ServerVM::isClassVisible(TR_OpaqueClassBlock *sourceClass, TR_OpaqueClassBl
 void *
 TR_J9ServerVM::getJ2IThunk(char *signatureChars, uint32_t signatureLength, TR::Compilation *comp)
    {
-   // Return NULL, since thunk will be recreated on the server whether or not it already exists.
-   // TODO: Track which thunks were created per-client, so that we don't need to recreate them.
+   // Check a set of thunks already registered (or about to be registered) for this client.
+   // If a thunk is there, return -1 to indicate that we don't need to regenerate thunk on
+   // the server and send it to client again. -1 is fine here, because the address will be relocated
+   // on the client anyway.
+   // If a thunk is not registered, return NULL to indicate that we need to generate and register it.
+      {
+      OMR::CriticalSection thunkMonitor(_compInfoPT->getClientData()->getThunkSetMonitor());
+      auto &thunkSet = _compInfoPT->getClientData()->getRegisteredJ2IThunkSet();
+      std::string signature(signatureChars, signatureLength);
+      if (thunkSet.find(signature) != thunkSet.end())
+         return reinterpret_cast<void *>(-1);
+      }
    return NULL;
    }
 
 void *
 TR_J9ServerVM::setJ2IThunk(char *signatureChars, uint32_t signatureLength, void *thunkptr, TR::Compilation *comp)
    {
-   TR_J9VMBase::setJ2IThunk(signatureChars, signatureLength, thunkptr, comp);
+   // Serialize thunk and send it to the client to be registered with the VM.
```

Comment:
A potentially problematic situation due to a race:
1) Thread 1 at server does `setJ2IThunk`, registers the thunk locally in the set and sends a message to the client
2) The client keeps this thunk in its local list, to be registered at the very end of the compilation
3) Thread 2 at the server does `getJ2IThunk`, finds the thunk in the set, continues and finishes the compilation
4) The client tries to relocate the thunk for thread 2, searches the VM, but there is nothing to be found.

Reply:
I see the problem.
We could have a per-compilation cache of thunks that were generated but not registered by the client yet. At the end of the compilation, everything from per-compilation cache would be committed to persistent cache.
Then `getJ2IThunk` could access only the persistent cache.